scikeras/wrappers.py

```diff
@@ -394,46 +386,55 @@ def _fit_keras_model(self, X, y, sample_weight, warm_start):
         else:
             hist = self.model_.fit(x=X, y=y, **fit_args)
 
-        if warm_start:
-            if not hasattr(self, "history_"):
-                self.history_ = defaultdict(list)
-            self.history_ = {
-                get_metric_full_name(k): self.history_[get_metric_full_name(k)]
-                + hist.history[k]
-                for k in hist.history.keys()
-            }
-        else:
-            self.history_ = hist.history
-        self.is_fitted_ = True
+        if not warm_start or not hasattr(self, "history_"):
+            self.history_ = defaultdict(list)
+
+        for key, val in hist.history.items():
+            try:
+                key = metric_name(key)
+            except ValueError as e:
+                # Keras puts keys like "val_accuracy" and "loss" and
+                # "val_loss" in hist.history.
+                if "Unknown metric function" not in str(e):
+                    raise e
+            self.history_[key] += [val]
 
         # return self to allow fit_transform and such to work
         return self
 
-    def _check_output_model_compatibility(self, y):
-        """Checks that the model output number and y shape match, reshape as needed.
+    def _check_model_compatibility(self, y: np.ndarray) -> None:
+        """Checks that the model output number and y shape match.
 
-        This is mainly in place to avoid cryptic TF errors.
+        This is in place to avoid cryptic TF errors.
         """
         # check if this is a multi-output model
-        if self.model_n_outputs_ != len(self.model_.outputs):
-            raise RuntimeError(
-                "Detected an input of size "
-                "{}, but {} has {} outputs".format(
-                    (y[0].shape[0], len(y)), self.model_, len(self.model_.outputs),
-                )
+        if self.n_outputs_expected_ != len(self.model_.outputs):
+            raise ValueError(
+                "Detected a Keras model input of size"
+                f" {y[0].shape[0]}, but {self.model_} has"
+                f" {self.model_.outputs} outputs"
             )
+        # check that if the user gave us a loss function it ended up in
+        # the actual model
+        default_val = inspect.signature(self.__init__).parameters["loss"].default
```

Comment:
This fails for any subclassed model that does not accept the `loss` parameter *directly*:
```python3
class MLPClassifier(KerasClassifier):

    def __init__(
        self,
        hidden_layer_sizes=(100, ),
        optimizer="adam",
        optimizer__learning_rate=0.001,
        epochs=200,
        verbose=0,
    ):
        super().__init__()
```
```
KeyError: 'loss'
```

Reply:
@stsievert I think we can catch the KeyError and return immediately (i.e. don't do any checks). The alternative would be to traverse the MRO or something; that seems like a lot of added complexity to me. What do you think?

Regardless, we should add a test that checks that basic subclassed classifiers with _no_ params work:
```python3
class MLPClassifier(KerasClassifier):
    def __init__(self):
        super().__init__()
```

Reply:
I see the problem. I think the if-statement you added is good; I added one small style nit.